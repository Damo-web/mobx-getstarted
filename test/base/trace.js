"use strict"

const mobx = require("../../src/mobx.ts")

test("trace", () => {
    mobx._resetGlobalState()
    const baselog = console.log
    try {
        const lines = []
        console.log = function() {
            lines.push(arguments)
        }

        const x = mobx.observable({
            firstname: "Michel",
            lastname: "Weststrate",
            get fullname() {
                var res = this.firstname + " " + this.lastname
                mobx.trace(this, "fullname")
                return res
            }
        })

        lines.push("- INIT -")
        x.fullname

        lines.push("- SECOND READ -")
        x.fullname

        lines.push("- REACTION -")

        const d = mobx.autorun(
            d => {
                x.fullname
                d.trace()
            },
            { name: "loggerzz" }
        )

        lines.push("- CHANGE -")

        mobx.transaction(() => {
            x.firstname = "John"
            x.lastname = "Doe"
        })

        lines.push("- DISPOSE -")

        d()

        expect(lines).toMatchSnapshot()
    } finally {
        console.log = baselog
    }
})
