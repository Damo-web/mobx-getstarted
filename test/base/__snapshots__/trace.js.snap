// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trace 1`] = `
Array [
  "- INIT -",
  Arguments [
    "[mobx.trace] 'ObservableObject@1.fullname' tracing enabled",
  ],
  "- SECOND READ -",
  Arguments [
    "[mobx.trace] 'ObservableObject@1.fullname' is being read outside a reactive context. Doing a full recompute",
  ],
  "- REACTION -",
  Arguments [
    "[mobx.trace] 'loggerzz' tracing enabled",
  ],
  "- CHANGE -",
  Arguments [
    "[mobx.trace] 'ObservableObject@1.fullname' is invalidated due to a change in: 'ObservableObject@1.firstname'",
  ],
  Arguments [
    "[mobx.trace] 'loggerzz' is invalidated due to a change in: 'ObservableObject@1.fullname'",
  ],
  "- DISPOSE -",
]
`;
